### ISSUE TITLE
"* Closing connection 0" on all requests

### ISSUE THREAD

**Issue by remy** [Orange-OpenSource/hurl#377]
I suspect this is related somehow to libcurl, but on every request I make I get `* Closing connection 0` on stderr.

I'm on macOS Big Sur, 11.5.2 (20G95). hurl is a new installation (1.4.0). Are there any tips on how I would solve this?

I've noticed that running a plain curl on the url doesn't show the same `* Closing connection 0`.

**Comment by jcamiel** [comment]
Humm... I'm testing Hurl on Mojave so don't I'm not surprised if there is some difference.

Could you please type this command in a shell:

```shell
which curl
otool -L [YOUR-PATH-TO-CURL-BINARY]
which hurl
otool -L [YOUR-PATH-TO-HURL-BINARY]
```

I would like to be sure which curl and lib curl is used (if it's system `curl` or Homebrew `curl`).

If you can, would it be possible to do a rebuild of Hurl and test if you still have the problem:
- install rust, either with `brew` or simply 
`curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh` from <https://www.rust-lang.org/learn/get-started>.
- clone the Hurl repo,
- in the repo, run `cargo install hurl`

**Comment by remy** [comment]
Just initially posting the otool bits you asked for - I'm going to get on with the rust install and see if I can build from scratch too and test.

```
$ otool -L /usr/bin/curl
/usr/bin/curl:
	/usr/lib/libcurl.4.dylib (compatibility version 7.0.0, current version 9.0.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.120.1)

$ /usr/local/bin/hurl:
	/usr/lib/libcurl.4.dylib (compatibility version 7.0.0, current version 9.0.0)
	/usr/lib/libxml2.2.dylib (compatibility version 10.0.0, current version 10.9.0)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)
	/usr/lib/libresolv.9.dylib (compatibility version 1.0.0, current version 1.0.0)
```

**Comment by jcamiel** [comment]
Thanks, it's very helpfull, I wanted to be sure that `curl` and `hurl` use the same system `libcurl`. I definitively need to try to reproduce it on Big Sur.

**Comment by remy** [comment]
Freshly installed from repo and same thing:

```
/Users/remy/.cargo/bin/hurl ~/Desktop/test.hurl
error: Assert Status
  --> /Users/remy/Desktop/test.hurl:4:8
   |
 4 | HTTP/* 201
   |        ^^^ actual value is <200>
   |

* Closing connection 0
```

**Comment by jcamiel** [comment]
Thanks a lot for your tests @remy I'm going to investigate it asap. 

Meanwhile, as a patch, you can redirect stderr to `/dev/null`($ hurl 2>/dev/null). I will update the issue with what I found.

**Comment by remy** [comment]
Cheers - though the work around does also swallow the error reports for legit errors (like the assert failure I posted previously gets swallowed). I can live with a line on stderr for the time being :)

**Comment by jcamiel** [comment]
I've reproduced the bug with a fresh macOS Big Sur.

The bug is in the system libcurl, which is in version 7.64.1. In libcurl 7.64.1 there is a bug <https://github.com/curl/curl/issues/3891> where some messages are wrongly outputed to stderr.
Fortunately, this bug has been fixed with curl/libcurl 7.65.0.

A simple workaround is to use [a more recent version of libcurl](https://hurl.dev/docs/frequently-asked-questions.html#how-can-i-use-a-custom-libcurl-from-homebrew-by-instance). With Homebrew this is really easy!

1. First, check that hurl is currently using the system libcurl:

```shell
$ otool -L /usr/local/bin/hurl:
	/usr/lib/libcurl.4.dylib (compatibility version 7.0.0, current version 9.0.0)
	/usr/lib/libxml2.2.dylib (compatibility version 10.0.0, current version 10.9.0)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)
	/usr/lib/libresolv.9.dylib (compatibility version 1.0.0, current version 1.0.0)
```
2. Then install a recent version of libcurl:

```shell
$ brew install curl
```

3. Then, update hurl binary to use the brew libcurl version:

```shell
$ sudo install_name_tool -change /usr/lib/libcurl.4.dylib /usr/local/opt/curl/lib/libcurl.4.dylib /usr/local/bin/hurl
```

4. Check that hurl is linked to the newly libcurl:

```shell
$ otool -L /usr/local/bin/hurl:
	/usr/local/opt/curl/lib/libcurl.4.dylib (compatibility version 7.0.0, current version 9.0.0)
	/usr/lib/libxml2.2.dylib (compatibility version 10.0.0, current version 10.9.0)
	/usr/lib/libiconv.2.dylib (compatibility version 7.0.0, current version 7.0.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1281.100.1)
	/usr/lib/libresolv.9.dylib (compatibility version 1.0.0, current version 1.0.0)
```

5. Enjoy Hurl without erroneous stderr !

I will update the [macOS Hurl FAQ](https://hurl.dev/docs/frequently-asked-questions.html#macos) with this bug on Big Sur.

Thanks for your issue report!

**Comment by jcamiel** [comment]
Updated docs [here](https://hurl.dev/docs/frequently-asked-questions.html#hurl-warning-on-big-sur-closing-connection-0)
